hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFreezeAccountHandler.java

```diff
@@ -44,23 +54,99 @@ public TokenFreezeAccountHandler() {
     public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
         requireNonNull(context);
         final var op = context.body().tokenFreezeOrThrow();
+        pureChecks(op);
+
         final var tokenStore = context.createStore(ReadableTokenStore.class);
         final var tokenMeta = tokenStore.getTokenMeta(op.tokenOrElse(TokenID.DEFAULT));
         if (tokenMeta == null) throw new PreCheckException(INVALID_TOKEN_ID);
         if (tokenMeta.hasFreezeKey()) {
             context.requireKey(tokenMeta.freezeKey());
+        } else {
+            throw new PreCheckException(TOKEN_HAS_NO_FREEZE_KEY);
         }
     }
 
     /**
      * This method is called during the handle workflow. It executes the actual transaction.
      *
-     * <p>Please note: the method signature is just a placeholder which is most likely going to
-     * change.
-     *
+     * @param txn the {@link TokenFreezeAccountTransactionBody} of the active transaction
+     * @param accountStore the {@link ReadableAccountStore} for the active transaction
+     * @param tokenStore the {@link ReadableTokenStore} for the active transaction
+     * @param tokenRelStore the {@link WritableTokenRelationStore} for the active transaction
      * @throws NullPointerException if one of the arguments is {@code null}
      */
-    public void handle() {
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull ReadableAccountStore accountStore,
+            @NonNull ReadableTokenStore tokenStore,
+            @NonNull WritableTokenRelationStore tokenRelStore)
+            throws HandleException {
+        requireNonNull(txn);
+        requireNonNull(accountStore);
+        requireNonNull(tokenStore);
+        requireNonNull(tokenRelStore);
+
+        final var op = txn.tokenFreezeOrThrow();
+        final var tokenRel = validateSemantics(op, accountStore, tokenStore, tokenRelStore);
+
+        final var copyBuilder = tokenRel.copyBuilder();
+        copyBuilder.frozen(true);
```

Comment:
seems we are missing all this from existing logic 🤔 
```
private TokenRelationship buildTokenRelationship(
            final Token token, final Account account, final HederaTokenRel merkleTokenRel) {
        final var tokenRelationship = new TokenRelationship(token, account);
        tokenRelationship.initBalance(merkleTokenRel.getBalance());
        tokenRelationship.setKycGranted(merkleTokenRel.isKycGranted());
        tokenRelationship.setFrozen(merkleTokenRel.isFrozen());
        tokenRelationship.setAutomaticAssociation(merkleTokenRel.isAutomaticAssociation());
        tokenRelationship.markAsPersisted();

        return tokenRelationship;
    }
```

Reply:
Ah good call, I didn't look at `ReadOnlyTokenStore.loadTokenRelationship`! Thanks! 🙇

Reply:
Don't think this is needed if we are suing `getForModify` 🤔

Reply:
@tinker-michaelj confirmed that this is only a "...part of trying to introduce a store abstraction with model objects," so we don't need to recreate